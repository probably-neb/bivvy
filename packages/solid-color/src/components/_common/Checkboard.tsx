import * as checkboard from '../../helpers/checkboard'
import { JSX, mergeProps } from 'solid-js'

export type CheckboardProps = {
  size?: number
  white?: string
  grey?: string
  renderers?: any
  borderRadius?: string | number
  boxShadow?: string
  children?: JSX.Element
}

export function Checkboard(_props: CheckboardProps) {
  const props = mergeProps(
    {
      white: 'transparent',
      grey: 'rgba(0,0,0,.08)',
      size: 8,
      renderers: {},
    },
    _props,
  )

  const styles = () => {
    const { size, white, grey, borderRadius, boxShadow, renderers } = props
    return {
      grid: {
        'border-radius': borderRadius,
        'box-shadow': boxShadow,
        position: 'absolute',
        inset: '0px',
        background: `url(${checkboard.get(white, grey, size, renderers.canvas)}) center left`,
      } as JSX.CSSProperties,
    }
  }

  // return isValidElement(children) ? (
  //   React.cloneElement(children, {
  //     ...children.props,
  //     style: { ...children.props.style, ...styles.grid },
  //   })
  // ) : (
  //   <div style={styles.grid} />
  // )
  // 判断children是否是一个有效的元素
  return props.children ? (
    // clone
    <div style={styles().grid}>{props.children}</div>
  ) : (
    <div style={styles().grid} />
  )
}
