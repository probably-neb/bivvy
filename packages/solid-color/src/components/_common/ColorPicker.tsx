import {
  Accessor,
  Context,
  createContext,
  createEffect,
  createMemo,
  createSignal,
  JSX,
  mergeProps,
  useContext,
} from 'solid-js'
import { ChangeColor, Color, ColorResult } from '../../types'
import * as color from '../../helpers/color'
import { debounce } from 'lodash-es'

export interface ColorPickerContextType {
  colors: Accessor<ColorResult>
  changeColor: (color: ChangeColor, event?: Event) => void
  onSwatchHover?: (color: ChangeColor, event: Event) => void
}

export const ColorPickerContext = createContext<ColorPickerContextType | undefined>(undefined)

export interface ColorPickerProps {
  children?: JSX.Element
  defaultColor?: Color
  color?: Color
  onChange?: (color: ColorResult, event?: Event) => void
  onChangeComplete?: (color: ColorResult) => void
  onSwatchHover?: (color: ColorResult, event: Event) => void
}

export function ColorPickerProvider(_props: ColorPickerProps) {
  const props = mergeProps({ defaultColor: { h: 250, s: 0.5, l: 0.2, a: 1 } }, _props)

  const [colors, setColors] = createSignal<ColorResult>({
    ...color.toState(props.color ?? props.defaultColor, 0),
  })

  createEffect(() => {
    if (props.color) {
      setColors({ ...color.toState(props.color, 0) })
    }
  })

  const handler = (fn: any, data: any, event: any) => fn(data, event)
  const debouncedChangeHandler = createMemo(() => debounce(handler, 100), [])

  const changeColor = (newColor: ChangeColor, event?: Event) => {
    const isValidColor = color.simpleCheckForValidColor(newColor)
    if (isValidColor) {
      const newColors = color.toState(
        newColor,
        (typeof newColor !== 'string' && 'h' in newColor ? newColor.h : undefined) ||
          colors().oldHue,
      )

      setColors(newColors)

      props.onChangeComplete && debouncedChangeHandler()(props.onChangeComplete, newColors, event)
      props.onChange && props.onChange(newColors, event)
    }
  }

  const handleSwatchHover = (data: ChangeColor, event: Event) => {
    const isValidColor = color.simpleCheckForValidColor(data)
    if (isValidColor) {
      const newColors = color.toState(
        data,
        (typeof data !== 'string' && 'h' in data ? data.h : undefined) || colors().oldHue,
      )
      props.onSwatchHover && props.onSwatchHover(newColors, event)
    }
  }

  const store = {
    colors,
    changeColor,
    onSwatchHover: props.onSwatchHover ? handleSwatchHover : undefined,
  }

  return <ColorPickerContext.Provider value={store}>{props.children}</ColorPickerContext.Provider>
}

export function useColorPicker() {
  return useContext(ColorPickerContext as Context<ColorPickerContextType>)
}

export function withColorPicker<T extends object>(Component: (props: T) => JSX.Element) {
  return (props: T & Omit<ColorPickerProps, 'children'>) => (
    <ColorPickerProvider {...props}>
      <Component {...props} />
    </ColorPickerProvider>
  )
}
