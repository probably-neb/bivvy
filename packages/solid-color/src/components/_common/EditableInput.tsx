import { merge } from 'lodash-es'
import { createEffect, createSignal, JSX, mergeProps, onCleanup, Show } from 'solid-js'

export interface EditableInputProps {
  styles: Record<string, JSX.CSSProperties>
  value?: string | number
  label?: string
  hideLabel?: boolean
  placeholder?: string
  arrowOffset?: number
  dragLabel?: boolean
  dragMax?: number
  onChange?: (value: any, e: Event) => void
}

interface EditableInputState {
  value: string
  blurValue: string
  borderColor?: string
}

export function EditableInput(_props: EditableInputProps) {
  const props = mergeProps(
    {
      arrowOffset: 1,
      hideLabel: false,
    },
    _props,
  )

  let inputRef: HTMLInputElement

  const inputId = `sc-editable-input-${Math.random().toString().slice(2, 5)}`

  const [state, setState] = createSignal<EditableInputState>({
    value: String(props.value).toUpperCase(),
    blurValue: String(props.value).toUpperCase(),
  })

  const DEFAULT_ARROW_OFFSET = 1
  const UP_KEY_CODE = 38
  const DOWN_KEY_CODE = 40
  const VALID_KEY_CODES = [UP_KEY_CODE, DOWN_KEY_CODE]
  const isValidKeyCode = (keyCode: number) => VALID_KEY_CODES.indexOf(keyCode) > -1
  const getNumberValue = (value: string) => Number(String(value).replace(/%/g, ''))

  const getValueObjectWithLabel = (value: string) => {
    return {
      [props.label!]: value,
    }
  }

  const setUpdatedValue = (value: any, e: Event) => {
    const onChangeValue = props.label ? getValueObjectWithLabel(value) : value
    props.onChange && props.onChange(onChangeValue, e)
    setState({ value, blurValue: value })
  }

  const handleBlur = () => {
    if (state().blurValue) {
      setState({ value: state().blurValue, blurValue: '' })
    }
  }

  const handleChange = (e: Event) => {
    setUpdatedValue((e.target as HTMLInputElement).value, e)
  }

  const handleDrag = (e: MouseEvent) => {
    if (props.dragLabel) {
      const newValue = Math.round(+props.value! + e.movementX)
      if (newValue >= 0 && newValue <= props.dragMax!) {
        props.onChange && props.onChange(getValueObjectWithLabel(String(newValue)), e)
      }
    }
  }

  const unbindEventListeners = () => {
    window.removeEventListener('mousemove', handleDrag)
    window.removeEventListener('mouseup', handleMouseUp)
  }

  const handleMouseUp = () => {
    unbindEventListeners()
  }

  const handleMouseDown = (e: MouseEvent) => {
    if (props.dragLabel) {
      e.preventDefault()
      handleDrag(e)
      window.addEventListener('mousemove', handleDrag)
      window.addEventListener('mouseup', handleMouseUp)
    }
  }

  const handleKeyDown = (e: KeyboardEvent) => {
    const value = getNumberValue((e.target as HTMLInputElement).value)
    if (!isNaN(value) && isValidKeyCode(e.keyCode)) {
      const offset = props.arrowOffset || DEFAULT_ARROW_OFFSET
      const updatedValue = e.keyCode === UP_KEY_CODE ? value + offset : value - offset
      setUpdatedValue(updatedValue, e)
    }
  }

  createEffect(() => {
    setState({
      value: String(props.value).toUpperCase(),
      blurValue: '',
    })
  })

  const styles = () => {
    return merge<Record<string, JSX.CSSProperties>, Record<string, JSX.CSSProperties>>(
      {
        wrap: {
          position: 'relative',
        },
      },
      props.styles,
    )
  }

  onCleanup(() => unbindEventListeners())

  return (
    <div style={styles().wrap}>
      <input
        id={inputId}
        ref={inputRef!}
        style={styles().input}
        spellcheck={false}
        value={state().value}
        placeholder={props.placeholder}
        onBlur={handleBlur}
        onChange={handleChange}
        onKeyDown={handleKeyDown}
        onInput={handleChange}
      />
      <Show when={props.label && !props.hideLabel}>
        <label style={styles().label} onMouseDown={handleMouseDown}>
          {props.label}
        </label>
      </Show>
    </div>
  )
}
