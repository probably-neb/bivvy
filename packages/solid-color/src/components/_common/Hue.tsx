import { JSX, mergeProps, onCleanup } from 'solid-js'
import { calculateChange } from '../../helpers/hue'
import { HslColor } from '../../types'

interface HueProps {
  children?: JSX.Element
  direction?: string
  radius?: number | string
  shadow?: string
  hsl: HslColor
  styles?: Record<string, JSX.CSSProperties>
  pointer?: <T extends object>(props: T) => JSX.Element
  onChange?: (data: HslColor, e: MouseEvent) => void
}

export function Hue(_props: HueProps) {
  const props = mergeProps(
    {
      direction: 'horizontal',
    },
    _props,
  )

  let container: HTMLDivElement

  const styles = () => {
    return {
      hue: {
        position: 'absolute',
        inset: '0px',
        'border-radius': typeof props.radius === 'string' ? props.radius : `${props.radius}px`,
        'box-shadow': props.shadow,
      },
      container: {
        padding: '0 2px',
        position: 'relative',
        height: '100%',
        'border-radius': typeof props.radius === 'string' ? props.radius : `${props.radius}px`,
      },
      pointer: {
        position: 'absolute',
        left: props.direction === 'vertical' ? '0px' : `${(props.hsl.h * 100) / 360}%`,
        top: props.direction === 'vertical' ? `${-((props.hsl.h * 100) / 360) + 100}%` : undefined,
      },
      slider: {
        'margin-top': '1px',
        width: '4px',
        'border-radius': '1px',
        height: '8px',
        'box-shadow': '0 0 2px rgba(0, 0, 0, .6)',
        background: '#fff',
        transform: 'translateX(-2px)',
      },
    } as Record<string, JSX.CSSProperties>
  }

  const handleChange = (e: MouseEvent) => {
    const change = calculateChange(e, props.direction, props.hsl, container)
    change && typeof props.onChange === 'function' && props.onChange(change, e)
  }

  const unbindEventListeners = () => {
    window.removeEventListener('mousemove', handleChange)
    window.removeEventListener('mouseup', handleMouseUp)
  }

  const handleMouseUp = () => {
    unbindEventListeners()
  }

  const handleMouseDown = (e: MouseEvent) => {
    handleChange(e)
    window.addEventListener('mousemove', handleChange)
    window.addEventListener('mouseup', handleMouseUp)
  }

  onCleanup(() => unbindEventListeners())

  return (
    <div style={styles().hue}>
      <div
        ref={container!}
        class={`hue-${props.direction}`}
        style={styles().container}
        onMouseDown={handleMouseDown}
      >
        <style>{`
          .hue-horizontal {
            background: linear-gradient(to right, #f00 0%, #ff0 17%, #0f0
              33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
            background: -webkit-linear-gradient(to right, #f00 0%, #ff0
              17%, #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
          }
          .hue-vertical {
            background: linear-gradient(to top, #f00 0%, #ff0 17%, #0f0 33%,
              #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
            background: -webkit-linear-gradient(to top, #f00 0%, #ff0 17%,
              #0f0 33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);
          }
        `}</style>
        <div style={styles().pointer}>
          {props.pointer ? <props.pointer {...props} /> : <div style={styles().slider} />}
        </div>
      </div>
    </div>
  )
}
