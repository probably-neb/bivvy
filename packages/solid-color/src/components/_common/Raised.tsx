import { merge } from 'lodash-es'
import { JSX, mergeProps } from 'solid-js'

export interface RaisedProps {
  background?: string
  zDepth?: 0 | 1 | 2 | 3 | 4 | 5
  radius?: number
  styles?: Record<string, JSX.CSSProperties>
  children?: JSX.Element
}

export function Raised(_props: RaisedProps) {
  const props = mergeProps(
    {
      zDepth: 1,
      radius: 2,
      background: '#fff',
      styles: {},
    },
    _props,
  )
  const styles = merge<Record<string, JSX.CSSProperties>, Record<string, JSX.CSSProperties>>(
    {
      wrap: {
        position: 'relative',
        display: 'inline-block',
      },
      content: {
        position: 'relative',
      },
      bg: {
        position: 'absolute',
        inset: '0px',
        'box-shadow':
          props.zDepth === 1
            ? '0 2px 10px rgba(0,0,0,.12), 0 2px 5px rgba(0,0,0,.16)'
            : `0 ${props.zDepth}px ${props.zDepth * 4}px rgba(0,0,0,.24)`,
        'border-radius': props.radius,
        background: props.background,
      },
    },
    props.styles,
  )

  return (
    <div style={styles.wrap}>
      <div style={styles.bg} />
      <div style={styles.content}>{props.children}</div>
    </div>
  )
}
