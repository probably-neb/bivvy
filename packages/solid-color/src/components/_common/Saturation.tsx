import { merge } from 'lodash-es'
import { createEffect, JSX, mergeProps } from 'solid-js'
import { calculateChange } from '../../helpers/saturation'
import { ChangeColor, HslColor, HsvColor } from '../../types'

export type SaturationProps = {
  hsl: HslColor
  hsv: HsvColor
  pointer?: JSX.Element
  onChange?: (color: ChangeColor, event?: Event) => void
  shadow?: JSX.CSSProperties['box-shadow']
  radius?: JSX.CSSProperties['border-radius']
  styles?: Record<string, JSX.CSSProperties>
}

export function Saturation(_props: SaturationProps) {
  const props = mergeProps({ styles: {} }, _props)
  let container: HTMLDivElement

  createEffect(() => {
    return () => {
      unbindEventListeners()
    }
  }, [])

  function handleChange(event: MouseEvent | TouchEvent) {
    if (props.onChange) {
      props.onChange(calculateChange(event, props.hsl, container), event as any)
    }
  }

  function handleMouseDown(event: MouseEvent) {
    handleChange(event)

    if (container) {
      container.addEventListener('mousemove', handleChange)
      container.addEventListener('mouseup', handleMouseUp)
    }
  }

  function handleMouseUp() {
    unbindEventListeners()
  }

  function unbindEventListeners() {
    if (container) {
      container.removeEventListener('mousemove', handleChange)
      container.removeEventListener('mouseup', handleMouseUp)
    }
  }

  const styles = () => {
    const { hsv, hsl, shadow, radius, styles } = props
    return merge<Record<string, JSX.CSSProperties>, Record<string, JSX.CSSProperties>>(
      {
        color: {
          position: 'absolute',
          inset: '0px',
          background: `hsl(${hsl.h},100%, 50%)`,
          'border-radius': radius,
        },
        white: {
          position: 'absolute',
          inset: '0px',
          'border-radius': radius,
        },
        black: {
          position: 'absolute',
          inset: '0px',
          boxShadow: shadow,
          'border-radius': radius,
        },
        pointer: {
          position: 'absolute',
          top: `${-(hsv.v * 100) + 100}%`,
          left: `${hsv.s * 100}%`,
          cursor: 'default',
        },
        circle: {
          width: '4px',
          height: '4px',
          'box-shadow': `0 0 0 1.5px #fff, inset 0 0 1px 1px rgba(0,0,0,.3),
            0 0 1px 2px rgba(0,0,0,.4)`,
          'border-radius': '50%',
          cursor: 'hand',
          transform: 'translate(-2px, -2px)',
        },
      },
      styles,
    )
  }

  return (
    <div
      style={styles().color}
      ref={container!}
      onMouseDown={handleMouseDown}
      onTouchMove={handleChange}
      onTouchStart={handleChange}
    >
      <style>{`
          .saturation-white {
            background: -webkit-linear-gradient(to right, #fff, rgba(255,255,255,0));
            background: linear-gradient(to right, #fff, rgba(255,255,255,0));
          }
          .saturation-black {
            background: -webkit-linear-gradient(to top, #000, rgba(0,0,0,0));
            background: linear-gradient(to top, #000, rgba(0,0,0,0));
          }
        `}</style>
      <div style={styles().white} class="saturation-white">
        <div style={styles().black} class="saturation-black" />
        <div style={styles().pointer}>
          {props.pointer ? props.pointer : <div style={styles().circle} />}
        </div>
      </div>
    </div>
  )
}
