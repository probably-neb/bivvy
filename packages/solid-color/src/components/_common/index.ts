export * from './Alpha'
export * from './EditableInput'
export * from './Raised'
export * from './Hue'
export * from './ColorPicker'
export * from './Checkboard'
export * from './Swatch'
export * from './Saturation'
export * from './Icons'
