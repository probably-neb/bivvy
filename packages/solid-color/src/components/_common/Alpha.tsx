import { merge } from 'lodash-es'
import { JSX, mergeProps, onCleanup } from 'solid-js'
import { HslColor, RgbColor } from '../../types'
import * as alpha from '../../helpers/alpha'
import { Checkboard } from './Checkboard'

export interface AlphaProps {
  rgb: RgbColor
  hsl: HslColor
  renderers?: any
  direction?: string
  a?: number
  radius?: number
  shadow?: string
  styles?: Record<string, JSX.CSSProperties>
  pointer?: <T extends object>(props: T) => JSX.Element
  onChange?: (data: any, e: Event) => void
}

export const Alpha = (_props: AlphaProps) => {
  const props = mergeProps(
    {
      direction: 'horizontal',
      styles: {},
    },
    _props,
  )
  let container: HTMLDivElement

  const styles = () => {
    const { rgb } = props
    return merge<Record<string, JSX.CSSProperties>, Record<string, JSX.CSSProperties>>(
      {
        alpha: {
          position: 'absolute',
          inset: '0px',
          'border-radius': props.radius,
        },
        checkboard: {
          position: 'absolute',
          inset: '0px',
          overflow: 'hidden',
          'border-radius': props.radius,
        },
        gradient: {
          position: 'absolute',
          inset: '0px',
          background:
            props.direction === 'vertical'
              ? `linear-gradient(to bottom, rgba(${rgb.r},${rgb.g},${rgb.b}, 0) 0%,
          rgba(${rgb.r},${rgb.g},${rgb.b}, 1) 100%)`
              : `linear-gradient(to right, rgba(${rgb.r},${rgb.g},${rgb.b}, 0) 0%,
         rgba(${rgb.r},${rgb.g},${rgb.b}, 1) 100%)`,
          'box-shadow': props.shadow,
          'border-radius': props.radius,
        },
        container: {
          position: 'relative',
          height: '100%',
          margin: '0 3px',
        },
        pointer: {
          position: 'absolute',
          left: props.direction === 'vertical' ? 0 : `${rgb.a && rgb.a * 100}%`,
          top: props.direction === 'vertical' ? `${rgb.a && rgb.a * 100}%` : undefined,
        },
        slider: {
          width: '4px',
          'border-radius': '1px',
          height: '8px',
          'box-shadow': '0 0 2px rgba(0, 0, 0, .6)',
          background: '#fff',
          'margin-top': '1px',
          transform: 'translateX(-2px)',
        },
      },
      props.styles,
    )
  }

  const handleChange = (e: Event) => {
    const change = alpha.calculateChange(e, props.hsl, props.direction, props.a, container)
    change && typeof props.onChange === 'function' && props.onChange(change, e)
  }

  const handleMouseDown = (e: Event) => {
    handleChange(e)
    window.addEventListener('mousemove', handleChange)
    window.addEventListener('mouseup', handleMouseUp)
  }

  const handleMouseUp = () => {
    unbindEventListeners()
  }

  const unbindEventListeners = () => {
    window.removeEventListener('mousemove', handleChange)
    window.removeEventListener('mouseup', handleMouseUp)
  }

  onCleanup(() => unbindEventListeners())

  return (
    <div style={styles().alpha}>
      <div style={styles().checkboard}>
        <Checkboard renderers={props.renderers} />
      </div>
      <div style={styles().gradient} />
      <div
        style={styles().container}
        ref={container!}
        onMouseDown={handleMouseDown}
        onMouseUp={handleChange}
        onTouchMove={handleChange}
        onTouchStart={handleChange}
      >
        <div style={styles().pointer}>
          {props.pointer ? <props.pointer {...props} /> : <div style={styles().slider} />}
        </div>
      </div>
    </div>
  )
}
