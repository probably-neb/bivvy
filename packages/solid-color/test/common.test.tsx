import { describe, it, expect } from 'vitest'
import { render } from '@solidjs/testing-library'
import { Alpha } from '../src/components/_common'
import { red } from '../src/helpers/color'
import '@testing-library/jest-dom'

describe('<Alpha />', () => {
  it('should render correctly', () => {
    const { container } = render(() => <Alpha {...red} />)
    expect(container).toMatchSnapshot()
  })
})
