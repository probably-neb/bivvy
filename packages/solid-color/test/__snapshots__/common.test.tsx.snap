// Vitest Snapshot v1

exports[`<Alpha /> > should render correctly 1`] = `
<div>
  <div
    style="position: absolute; inset: 0px;"
  >
    <div
      style="position: absolute; inset: 0px; overflow: hidden;"
    >
      <div
        style="position: absolute; inset: 0px; background: url(null) left;"
      />
    </div>
    <div
      style="position: absolute; inset: 0px;"
    />
    <div
      style="position: relative; height: 100%; margin: 0px 3px;"
    >
      <div
        style="position: absolute; left: 100%;"
      >
        <div
          style="width: 4px; border-radius: 1px; height: 8px; box-shadow: 0 0 2px rgba(0, 0, 0, .6); background: rgb(255, 255, 255); margin-top: 1px; transform: translateX(-2px);"
        />
      </div>
    </div>
  </div>
</div>
`;
